ansible_bender/callback_plugins/snapshoter.py

```diff
@@ -52,26 +56,57 @@ def _snapshot(self, task_result):
             return
-        content = self.get_task_content(task_result._task)
         if task_result.is_skipped() or getattr(task_result, "_result", {}).get("skip_reason", False):
-            a.record_progress(None, content, None, build_id=build.build_id)
+            a.record_progress(None, self._task_content, None, build_id=build.build_id)
             return
         # # alternatively, we can guess it's a file action and do getattr(task, "src")
         # # most of the time ansible says changed=True even when the file is the same
         if task_result._task.action in FILE_ACTIONS:
             if not task_result.is_changed():
-                status = a.maybe_load_from_cache(content, build_id=build.build_id)
+                status = a.maybe_load_from_cache(self._task_content, build_id=build.build_id)
                 if status:
                     self._display.display("loaded from cache: '%s'" % status)
                     return
-        image_name = a.cache_task_result(content, build)
+        image_name = a.cache_task_result(self._task_content, build)
         if image_name:
             self._display.display("caching the task result in an image '%s'" % image_name)
 
     @staticmethod
     def get_task_content(task: Task):
+
+        def get_templated_ds(task: Task):
+            # To get the correct fingerprint of a task,
+            # we need to do the jinja2 templating for its ds.
+            # The untemplated ds, if containing vars, will stay the same even if values of vars changed.
+            #
+            # We use ansible.template.Templar to help us do the templating,
+            # similar to what is done in ansible.executor.task_executor.
+            # However, the params needed by Templar (loader, shared_loader_obj, variables)
+            # are not easily-obtainable via Task object.
+            #
+            # The hacky solution to this is to use ansible.executor.process.worker.WorkerProcess obj.
+            # These params can be found in this obj.
+            # The WorkerProcess obj references the Task obj while running,
+            # so we can use gc.get_referrers method to find the WorkerProcess obj based on Task obj.
+            try:
+                required_fields = ["_loader", "_shared_loader_obj", "_task_vars"]
+                worker_process_dict = None
+                for referrer in gc.get_referrers(task):
+                    if isinstance(referrer, dict) and all([field in referrer for field in required_fields]):
+                        # Found the WorkerProcess obj referencing this Task obj.
+                        worker_process_dict = referrer
+                        break
+                templar = Templar(
+                    loader=worker_process_dict["_loader"],
+                    shared_loader_obj=worker_process_dict["_shared_loader_obj"],
+                    variables=worker_process_dict["_task_vars"],
+                )
+                return templar.template(task.get_ds())
```

Comment:
we should probably assert first that those keys are present in the dict so we wouldn't get KeyErrors in future if ansible-core ever changes its internals